### PR TITLE
feat: adds swapper registry

### PR DESCRIPTION
**Pull request by b0dhidharma**

### CODE REVIEW

File: contracts/SwapperRegistry.sol [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  mapping(address => string) public nameByAddress;
+  mapping(string => address) public swapperByName;
+  mapping(address => uint256) public initializationByAddress;
+  EnumerableSet.AddressSet internal _swappers;
+
+  mapping(address => EnumerableSet.AddressSet) internal _approvedTokensBySwappers;
+
+  constructor() {
+    // TODO: set governance
+    // TODO: set collectable dust
```
Comment (lbertenasco) [inline contracts/SwapperRegistry.sol:30]: You can use `UtilsReady` abstract contract from library

File: contracts/SwapperRegistry.sol [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  function _addSwapper(string memory _name, address _swapper) internal {
+    require(bytes(_name).length > 0, '');
+    require(_swapper != address(0), '');
+    require(!_swappers.contains(_swapper), '');
+    nameByAddress[_swapper] = _name;
+    swapperByName[_name] = _swapper;
+    initializationByAddress[_swapper] = block.timestamp;
+    _swappers.add(_swapper);
+    emit SwapperAdded();
+  }
```
Comment (lbertenasco) [inline contracts/SwapperRegistry.sol:68]: here check if name is not already in use.

File: contracts/SwapperRegistry.sol [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  function enableSwapper(string memory _name, address _token) external virtual {
+    // TODO: only governance or strategy
+    _enableSwapper(_name, _token);
+  }
+
+  function _enableSwapper(string memory _name, address _token) internal {
+    address _swapper = swapperByName[_name];
+    require(_swappers.contains(_swapper), '');
+    require(_token != address(0), '');
+    // TODO: maybe check if allowed and re-max it ?
```
Comment (lbertenasco) [inline contracts/SwapperRegistry.sol:92]: will never need a remax :)

File: contracts/SwapperRegistry.sol [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    _swappers.remove(_swapper);
+    emit SwapperRemoved();
+  }
+
+  function enableSwapper(string memory _name, address _token) external virtual {
+    // TODO: only governance or strategy
+    _enableSwapper(_name, _token);
+  }
+
+  function _enableSwapper(string memory _name, address _token) internal {
```
Comment (lbertenasco) [inline contracts/SwapperRegistry.sol:88]:
```suggestion
  function _enableSwapperToken(string memory _name, address _token) internal {
```
better naming?

File: contracts/SwapperRegistry.sol [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    address _swapper = swapperByName[_name];
+    require(_swappers.contains(_swapper), '');
+    require(_token != address(0), '');
+    // TODO: maybe check if allowed and re-max it ?
+    if (!_approvedTokensBySwappers[_swapper].contains(_token)) {
+      IERC20(_token).safeApprove(_swapper, type(uint256).max);
```
Comment (lbertenasco) [inline contracts/SwapperRegistry.sol:94]: :eyes: :heart_eyes:

File: contracts/SwapperRegistry.sol (hunk 1 above)
Comment (b0dhidharma) [reply contracts/SwapperRegistry.sol:30]: Can't it's in old pragma :sob: will probably import it and update the pragma.

File: contracts/SwapperRegistry.sol (hunk 2 above)
Comment (b0dhidharma) [reply contracts/SwapperRegistry.sol:68]: oof. good catch my dude.

File: contracts/SwapperRegistry.sol (hunk 3 above)
Comment (b0dhidharma) [reply contracts/SwapperRegistry.sol:92]: k. taken comment out then.

**Pull request merged**